contracts/test/StakeTable.t.sol

```diff
@@ -0,0 +1,225 @@
+
+        (
+            BN254.G2Point memory blsVK,
+            EdOnBN254.EdOnBN254Point memory schnorrVK,
+            BN254.G1Point memory sig
+        ) = computeGroupElements(msgSenderAddress);
+
+        // Prepare for the token transfer
+        vm.prank(msg.sender);
+        token.approve(address(stakeTable), depositAmount);
+
+        vm.prank(msgSenderAddress);
+
+        // Balances before registration
+        assertEq(token.balanceOf(msgSenderAddress), INITIAL_BALANCE);
+
+        uint256 nativeAmount;
+        uint256 restakedAmount;
+        (nativeAmount, restakedAmount) = stakeTable.totalStake();
+        assertEq(nativeAmount, 0);
+        assertEq(restakedAmount, 0);
+
+        // Check event is emitted
+        vm.expectEmit(false, false, false, true, address(stakeTable));
+        IStakeTable.Node memory node;
+        node.account = msgSenderAddress;
+        node.balance = depositAmount;
+        node.stakeType = IStakeTable.StakeType.Native;
+        node.schnorrVK = schnorrVK;
+        node.registerEpoch = 1;
+
+        // The function IStakeTable._hashBlsKey is private but its code is small hence we copy it
+        // there.
+        emit Registered(
+            keccak256(abi.encode(blsVK.x0, blsVK.x1, blsVK.y0, blsVK.y1)),
+            node.registerEpoch,
+            node.stakeType,
+            node.balance
+        );
+
+        // Successful call to register
+        vm.prank(msgSenderAddress);
+        bool res = stakeTable.register(
+            blsVK, schnorrVK, depositAmount, IStakeTable.StakeType.Native, sig, validUntilEpoch
+        );
+        assertTrue(res);
+
+        // Balance after registration
+        assertEq(token.balanceOf(msgSenderAddress), INITIAL_BALANCE - depositAmount);
+        (nativeAmount, restakedAmount) = stakeTable.totalStake();
+        assertEq(nativeAmount, depositAmount);
+        assertEq(restakedAmount, 0);
+    }
+
+    function testInvalidBLSSig() external {
+        (BN254.G2Point memory blsVK, EdOnBN254.EdOnBN254Point memory schnorrVK,) =
+            computeGroupElements(msg.sender);
+
+        // Failed signature verification
+        BN254.G1Point memory badSig = BN254.P1();
+        vm.expectRevert(BLSSig.BLSSigVerificationFailed.selector);
+        stakeTable.register(
+            blsVK, schnorrVK, depositAmount, IStakeTable.StakeType.Native, badSig, validUntilEpoch
+        );
+    }
+
+    function testRestakingNotImplemented() external {
+        (
+            BN254.G2Point memory blsVK,
+            EdOnBN254.EdOnBN254Point memory schnorrVK,
+            BN254.G1Point memory sig
+        ) = computeGroupElements(msg.sender);
+
+        // Throw "Restaking not implemented" error
+        vm.expectRevert(S.RestakingNotImplemented.selector);
+        stakeTable.register(
+            blsVK, schnorrVK, depositAmount, IStakeTable.StakeType.Restake, sig, validUntilEpoch
+        );
+    }
+
+    function testInvalidNextRegistrationEpoch() external {
```

Comment:
we should use `testFuzz` in this case, whenever we are testing some user-input, we should use fuzzing as much as we could to test for unexpected inputs.

namely this function should take in `uint64 validUntilEpoch` as an input (which will be filled by the foundry fuzzer), if you want to restrict its value/range. try to use [`bound()`](https://book.getfoundry.sh/reference/forge-std/bound). 
(sometimes you want rejection sampling using [`vm.assume()`](https://book.getfoundry.sh/cheatcodes/assume#assume), just FYI, also both of these have examples in my other tests, so you can take a look)

Reply:
Done in 44cfc9e.